greybox_generalize.py

```diff
@@ -0,0 +1,249 @@
+import numpy as np
+import pyomo.environ as pyo
+from scipy.sparse import coo_matrix
+from pyomo.contrib.pynumero.interfaces.external_grey_box import ExternalGreyBoxModel
+from pyomo.contrib.pynumero.interfaces.external_grey_box import ExternalGreyBoxBlock
+
+
+class LogDetModel(ExternalGreyBoxModel):
+    def __init__(
+        self,
+        n_parameters=2,
+        initial_fim=None,
+        use_exact_derivatives=True,
+        print_level=0,
+    ):
+        """
+        Greybox model to compute the log determinant of a sqaure symmetric matrix.
+
+        Arguments
+        ---------
+        n_parameters: int
+            Number of parameters in the model. The square symmetric matrix is of shape n_parameters*n_parameters
+        initial_fim: dict
+            Initial value of the matrix. If None, the identity matrix is used.
+        use_exact_derivatives: bool
+            If True, the exact derivatives are used.
+            If False, the finite difference approximation can be used, but not recommended/tested.
+        print_level: integer
+            0 (default): no extra output
+            1: minimal info to indicate if initialized well
+                print the following:
+                - initial FIM received by the grey-box moduel
+            2: intermediate info for debugging
+                print all the level 1 print statements, plus:
+                - the FIM output of the current iteration, both the output as the FIM matrix, and the flattened vector
+            3: all details for debugging
+                print all the level 2 print statements, plus:
+                - the log determinant of the FIM output of the current iteration
+                - the eigen values of the FIM output of the current iteration
+
+        Return
+        ------
+        None
+        """
+        self._use_exact_derivatives = use_exact_derivatives
+        self.print_level = print_level
+        self.n_parameters = n_parameters
+        self.num_input = int(
+            n_parameters + (n_parameters * n_parameters - n_parameters) // 2
+        )
+        self.initial_fim = initial_fim
+
+        # variable to store the output value
+        # Output constraint multiplier values. This is a 1-element vector because there is one output
+        self._output_con_mult_values = np.zeros(1)
+
+        if not use_exact_derivatives:
+            raise NotImplementedError("use_exact_derivatives == False not supported")
+
+    def input_names(self):
+        """Return the names of the inputs.
+        Define only the upper triangle of FIM because FIM is symmetric
+
+        Return
+        ------
+        input_name_list: a list of the names of inputs
+        """
+        # store the input names strings
```

Comment:
This says these are strings, but looking at the code below, these are integers.

Reply:
change to tuple